File: odroid/Control/src/bert/control/model/URDFModel.java
/**
 * Copyright 2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.control.model;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Iterator;
import java.util.logging.Logger;

import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import bert.share.control.Appendage;
import bert.share.motor.Joint;
import bert.share.xml.XMLUtility;

/**
 *  Parse the URDF file and make its contents available as a chain of links.
 *  A chain has a tree structure with a single root.
 */
public class URDFModel  {
	private static final String CLSS = "URDFModel";
	private static final Logger LOGGER = Logger.getLogger(CLSS);
	protected Document document;
	private final Chain chain;
	
	public URDFModel() {
		this.chain = new Chain();
		this.document = null;
	}
    
	/**
	 * @return the tree of links which describes the robot.
	 */
	public Chain getChain() { return this.chain; }

    /**
	 * Expand the supplied path as the URDF XML file.
	 * @return the geometry, an XML document.
	 */
	public void analyzePath(Path filePath) {
		LOGGER.info(String.format("%s.analyzePath: URDF file(%s)",CLSS,filePath.toAbsolutePath().toString()));
		try {
			byte[] bytes = Files.readAllBytes(filePath);
			if( bytes!=null ) {
				this.document = XMLUtility.documentFromBytes(bytes);
				analyzeChain();
			}
		}
		catch( IOException ioe) {
			LOGGER.severe(String.format("%s.analyzePath: Failed to read file %s (%s)",
											CLSS,filePath.toAbsolutePath().toString(),ioe.getLocalizedMessage()));
		}
	}

    // ================================ Auxiliary Methods  ===============================
	

	
	/**
	 * Search the model for link and joint elements. 
	 */
	private void analyzeChain() {
		if( this.document!=null ) {
			// ================================== IMU ===============================================
			NodeList imus = document.getElementsByTagName("imu");
			if( imus.getLength()>0 ) {  
				//LOGGER.info(String.format("%s.analyzeChain: IMU ...",CLSS));
				Node imuNode = imus.item(0); // Should only be one
				NodeList childNodes = imuNode.getChildNodes();
				int childCount = childNodes.getLength();
				int childIndex=0;
				String text = null;
				while(childIndex<childCount) {
					Node cNode = childNodes.item(childIndex);
					if( "origin".equalsIgnoreCase(cNode.getLocalName())) {
						text  = XMLUtility.attributeValue(cNode, "xyz");
						double[] xyz = doubleArrayFromString(text);
						chain.setOrigin(xyz);
					}
					else if( "axis".equalsIgnoreCase(cNode.getLocalName())) {
						text = XMLUtility.attributeValue(cNode, "xyz");
						double[] xyz = doubleArrayFromString(text);
						chain.setAxes(xyz);
					}
					childIndex++;
				}
			}
			
			// ================================== Links ===============================================
			NodeList links = document.getElementsByTagName("link");
			int count = links.getLength();
			int index = 0;
			while(index<count) {
				Node linkNode = links.item(index);
				String name = XMLUtility.attributeValue(linkNode, "name");
				LOGGER.info(String.format("%s.analyzeChain: Link %s ...",CLSS,name));
				try {
					chain.createLink(name.toUpperCase());

					NodeList appendages = linkNode.getChildNodes();
					int acount = appendages.getLength();
					int aindex=0;
					while(aindex<acount) {
						Node node = appendages.item(aindex);
						if( "appendage".equalsIgnoreCase(node.getLocalName())) {
							String aname = XMLUtility.attributeValue(node, "name");

							NodeList childNodes = node.getChildNodes();
							int childCount = childNodes.getLength();
							int childIndex=0;
							double[] xyz = null;
							double[] ijk = null;
							while(childIndex<childCount) {
								Node cNode = childNodes.item(childIndex);
								if( "origin".equalsIgnoreCase(cNode.getLocalName()))   xyz  = doubleArrayFromString(XMLUtility.attributeValue(cNode, "xyz"));
								else if("axis".equalsIgnoreCase(cNode.getLocalName())) ijk  = doubleArrayFromString(XMLUtility.attributeValue(cNode, "xyz"));
								childIndex++;
							}
							
							Appendage a = Appendage.valueOf(aname.toUpperCase());
							chain.createLink(a.name());
							LinkPoint end = new LinkPoint(a,ijk,xyz);
							chain.setEndPoint(a.name(),end);
							chain.setParent(a.name(),name.toUpperCase());
						}
						aindex++;
					}
				}
				catch(IllegalArgumentException iae) {
					LOGGER.warning(String.format("%s.analyzeChain: link or appendage has unknown name: %s, ignored (%s)",CLSS,name,iae.getLocalizedMessage()));
					iae.printStackTrace();
				}
				index++;
			}
			// ================================== Joints ===============================================
			NodeList joints = document.getElementsByTagName("joint");
			count = joints.getLength();
			index = 0;
			while(index<count) {
				Node jointNode = joints.item(index);
				String name = XMLUtility.attributeValue(jointNode, "name");
				LOGGER.info(String.format("%s.analyzeChain: Joint %s ...",CLSS,name));
				try {
					Joint joint = Joint.valueOf(name);
					NodeList childNodes = jointNode.getChildNodes();
					int childCount = childNodes.getLength();
					int childIndex = 0;
					String parent = null;
					String child = null;
					double[] xyz = null;
					double[] ijk = null;
					// It is required that the LinkPoint have a parent and a child
					while( childIndex<childCount ) {
						Node childNode = childNodes.item(childIndex);
						if( "parent".equals(childNode.getLocalName()) ) {
							parent = XMLUtility.attributeValue(childNode, "link");
							if( parent==null ) {
								LOGGER.warning(String.format("%s.analyzeChain: joint %s has no parent, ignored",CLSS,joint.name()));
							}
						}
						else if( "child".equals(childNode.getLocalName()) ) {
							child = XMLUtility.attributeValue(childNode, "link");
							if( child==null ) {
								
							}
						}
						else if( "origin".equalsIgnoreCase(childNode.getLocalName())) xyz  = doubleArrayFromString(XMLUtility.attributeValue(childNode, "xyz"));
						else if("axis".equalsIgnoreCase(childNode.getLocalName()))    ijk  = doubleArrayFromString(XMLUtility.attributeValue(childNode, "xyz"));
						childIndex++;
					}

					LinkPoint rev = new LinkPoint(joint,ijk,xyz);
					if(parent!=null) {
						Link parentLink = chain.getLinkForLimbName(parent);
						parentLink.setEndPoint(rev);
						if(child!=null ) {
							Link childLink = chain.getLinkForLimbName(child);
							chain.setOriginPoint(childLink,rev);
							childLink.setParent(parentLink);
						}
						else {
							LOGGER.warning(String.format("%s.analyzeChain: joint %s has no child",CLSS,joint.name()));
						}
					}
					else {
						LOGGER.warning(String.format("%s.analyzeChain: joint %s has no parent",CLSS,joint.name()));
					}
					
				}
				catch(IllegalArgumentException iae) {
					LOGGER.warning(String.format("%s.analyzeChains: link element has unknown name (%s), ignored",CLSS,name));
				}
				index++;
			}

			// Search for origin aka root. Choose any random link and follow to root.
			Iterator<Link> linkWalker = chain.getLinks().iterator();
			if( linkWalker.hasNext()) {
				Link link = linkWalker.next();
				while( link!=null ) {
					Link parent = link.getParent();
					if(parent==null) {
						chain.setRoot(link);
						break;
					}
					link = parent;
				}
			}
		}
	}
	
	// ============================================= Helper Methods ==============================================
	private double[] doubleArrayFromString(String text) {
		if( text==null ) return null; 
		double [] result = new double[3];
		String[] raw = text.split(" ");
		for(int i=0;i<raw.length;i++) {
			try {
				result[i] = Double.parseDouble(raw[1]);
			}
			catch(NumberFormatException nfe) {
				LOGGER.warning(String.format("%s.doubleArrayFromString: Error parsing %s (%s);",CLSS,text,nfe.getLocalizedMessage()));
			}
		}
		return result;
	}
}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/model/URDFModel.kt
/**
 * Copyright 2022-2026. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.common.model

import chuckcoughlin.bert.common.util.XMLUtility
import com.google.gson.GsonBuilder
import org.w3c.dom.Document
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.logging.Logger

/**
 * Parse the URDF file and make its contents available as a chain of links.
 * A chain has a tree structure with a single root.
 */
object URDFModel {
    private var document: Document?
    private val tree: JointTree

    /**
     * Expand the supplied path as the URDF XML file.
     * @return the geometry, an XML document.
     */
    fun analyzePath(filePath: Path) {
        LOGGER.info(String.format("%s.analyzePath: URDF file(%s)", CLSS, filePath.toAbsolutePath().toString()))
        try {
            val bytes = Files.readAllBytes(filePath)
            document = XMLUtility.documentFromBytes(bytes)
            analyzeChain()
        }
        catch (ioe: IOException) {
            LOGGER.severe(String.format("%s.analyzePath: Failed to read file %s (%s)",
                    CLSS, filePath.toAbsolutePath().toString(), ioe.getLocalizedMessage()))
        }
    }
    // ================================ Auxiliary Methods  ===============================
    /**
     * Search the model for IMU, link and joint elements.
     * IMU maps the location of the start of the chain with respect to the
     * center of gravity. The IMU reacts only to rotation.
     */
    private fun analyzeChain() {
        if (document != null) {
            var origin = JointPosition()
            origin.name = Joint.IMU.name
            origin.parent = ConfigurationConstants.NO_ID
            // ================================== IMU ===============================================
            val imus = document!!.getElementsByTagName("imu")
            if (imus.length > 0) {
                LOGGER.info(String.format("%s.analyzeChain: IMU ...",CLSS))
                val imuNode = imus.item(0) // Should only be one
                val xyz = doubleArrayFromString(XMLUtility.attributeValue(imuNode, "xyz"))
                origin.setCoordinates(xyz[0],xyz[1],xyz[2])
            }
            tree.setOrigin(origin)

            // ================================== Links ===============================================
            // Links are a connection between joints or from a joint to extremity (appendage). The link
            // that has no source is the IMU, the root if the tree.
            //
            // Create all the Links and their LinkPins. A LinkPin encompasses the connecting joint.
            // Create a separate link for each resolute joint and end effector.
            // The URDF file has all angles in degrees. Convert to radians.
            //
            val links = document!!.getElementsByTagName("link")
            val count = links.length
            var index = 0
            while (index < count) {
                val linkNode = links.item(index)
                try {
                    // The source is shared with all sub-links
                    var parent = origin
                    val children = linkNode.childNodes
                    val acount = children.length
                    var aindex = 0
                    while (aindex < acount) {
                        val node = children.item(aindex)
                        // Same source must be applied to all joints/appendages in same link element
                        if ("source".equals(node.localName)) {
                            val jname: String = XMLUtility.attributeValue(node, "joint")
                            parent = tree.getJointPositionByName(jname)
                        }
                        aindex++
                    }

                    aindex = 0
                    while (aindex < acount) {
                        val node = children.item(aindex)
                        if ("appendage".equals(node.localName) || "joint".equals(node.localName)) {
                            val aname: String = XMLUtility.attributeValue(node, "name")
                            val jp = tree.getJointPositionByName(aname)
                            var home = 0.0
                            if("appendage".equals(node.localName) ) {
                                jp.isAppendage = true
                            }
                            else  {
                                jp.isAppendage = false
                                home = XMLUtility.attributeValue(node, "home").toDouble()
                            }
                            jp.side = XMLUtility.attributeValue(linkNode, "side")
                            jp.parent = parent.id
                            var jlink = tree.createJointLink(parent,jp)
                            val rpy = doubleArrayFromString(XMLUtility.attributeValue(node, "rpy"))
                            jlink.setRpy(rpy[0],rpy[1],rpy[2])
                            val xyz = doubleArrayFromString(XMLUtility.attributeValue(node, "xyz"))
                            jlink.setDimensions(xyz[0],xyz[1],xyz[2])
                            jlink.sourceJoint.home = home
                        }
                        aindex++
                    }
                }
                catch (iae: IllegalArgumentException) {
                    LOGGER.warning(String.format("%s.analyzeChain: exception on link %d, ignored (%s)",
                        CLSS,index,iae.localizedMessage))
                    iae.printStackTrace()
                }
                index++
            }
        }
    }

    /**
     * Create a JointTree from the existing structure.
     * Note that all joint positions are (0,0,0).
     */
    fun createJointTree() : JointTree {
        val jtree = tree.clone()
        return jtree
    }



    /**
     * @return  a comma-separated string of the names of all extremities.
     */
    fun endEffectorNames(): String {
        val names = StringBuffer()
        for (jp in tree.listJointPositions()) {
            if(jp.isAppendage) {
                names.append(jp.name.lowercase())
                names.append(", ")
            }
        }
        if( names.isNotEmpty() ) return names.substring(0, names.length - 2)
        else return "none"
    }
    /**
     * @return  a JSON pretty-printed String array of all appendages.
     */
    fun endEffectorNamesToJSON(): String {
        val gson = GsonBuilder().setPrettyPrinting().create()
        val names = mutableListOf<String>()
        for (jp in tree.listJointPositions()) {
            if(jp.isAppendage) {
                names.add(jp.name)
            }
        }
        return gson.toJson(names)
    }

    /**
     * @return  a JSON pretty-printed String array of all property types. Exclude NONE.
     */
    fun jointsToJSON(): String {
        val gson = GsonBuilder().setPrettyPrinting().create()
        var names = mutableListOf<String>()
        for (jp in tree.listJointPositions()) {
            if(!jp.isAppendage) {
                names.add(jp.name)
            }
        }
        return gson.toJson(names)
    }
    // ============================================= Helper Methods ==============================================
    private fun doubleArrayFromString(text: String): DoubleArray {
        val result = DoubleArray(3)
        val raw = text.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (i in raw.indices) {
            try {
                result[i] = raw[i].toDouble()
            }
            catch (nfe: NumberFormatException) {
                LOGGER.warning(String.format("%s.doubleArrayFromString: Error parsing %s raw(%d)=%s (%s)",
                        CLSS,text,i,raw[i],nfe.localizedMessage ) )
            }
        }
        //if(DEBUG) LOGGER.info(String.format("%s.doubleArrayFromString: text %s = %.2f,%.2f,%.2f",CLSS,text,result[0],result[1],result[2]))
        return result
    }

    private val CLSS = "URDFModel"
    private val DEBUG: Boolean
    private val LOGGER = Logger.getLogger(CLSS)

    init {
        DEBUG= RobotModel.debug.contains(ConfigurationConstants.DEBUG_CONFIGURATION)
        document = null
        tree = JointTree()
    }
}